### DISCUSSION TITLE
Cursor based pagination

### DISCUSSION THREAD

**Discussion by bkniffler** [supabase/supabase#3938] [Questions] [answered]
Is cursor based pagination supported with supabase API?

So instead of giving a offset/limit, just be able to pass the last ID (cursor) and a limit?

**Comment by dshukertjr** [comment, marked as answer]
Hi @bkniffler!

What you are looking for is very much possible by utilizing filters like [`gt()`](https://supabase.io/docs/reference/javascript/gt) or [`lt()`](https://supabase.io/docs/reference/javascript/lt). `gt` stands for greater than and `lt` stands for less than. 

For example, if you are displaying posts in descending `created_at` order, you can get the next page like this:

```typescript
const { data, error } = await supabase
  .from('posts')
  .select()
  .lt('created_at', [LAST_POST_CREATED_AT])
  .order('created_at', { ascending: false })
  .limit(10)
```

**Reply by bkniffler** [reply]
Thanks, that's what I'm doing right now, and its working good!

**Reply by person2434** [reply]
Will there be an effect on the results if there is a change or deletion of the rows that were previously fetched?

**Reply by iriealex** [reply]
No, there will be no side effect.

However, there is a small problem with the proposed solution: if several posts have the same value for the `created_at` property, in the end this code would not work because some posts could be skipped!

**Reply by roker15** [reply]
> For example, if you are displaying posts in descending `created_at` order, you can get the next page like this:

I assume created_at is not indexed by default. Should this have impact on performance as compared to doing pagination based on primary key (which is id)?

**Reply by dshukertjr** [reply]
@roker15 For a large table, you may want to create an index on `created_at` because it will have an impact on performance as you have suspected!

**Reply by caseycrogers** [reply]
FWIW, this solution may still have major pitfalls depending on your data model and your order by. For example, I have a list of audio posts that are ordered by (roughly) the following logic:

1. `has_listened ASC`
2. `(created_at + virality_score) DESC`

`has_listened` is set by the client inserting a row into a table when the user listens to a post. The `virality_score` is fetched via a function. Because `virality_score` changes every time a post is interacted with and `has_listened` changes as the user listens through the query results, I'm not sure there's a coherent way for me not to get skipped or duplicated rows without cursors...open to any suggestions.

I could make my `virality_score` and `has_listened` both fetched via functions that take time as an input. They could then return the historical value at the given time. Then I could set this time to the start time of my paginated query to force the order by to be stable across pages. However, I'm not sure there's a way to coherently send an arbitrary client value to a server-defined view? Maybe I could do something hideous like issue an insert to a table that contains server-generated `query_id` and `start_time` and then have my view join against that? My client side logic would then look something like:
```dart
// This inserts a row to a table with a server generated timestamp and id.
final String queryId = getIdFromResult(queryIdTable.insert());

// Server side definition of view:
//  ```sql
//  select query_id,
//         query_start_time,
//         has_listened(query_start_time),
//         virality_score(query_start_time),
//         ...
//  from posts
//  full join query_ids
//  ```
idBasedPostsView.select()
  .eq('query_id' = queryId)
  .orderBy(...);
```

At this point I've basically just rebuilt cursors though...however the server state is just isolated to id:timestamp pairs which is extremely lightweight so this'll be a pain to maintain but is pretty scalable? Assuming the query optimizer is able to propagate the `eq('query_id', queryId)` downward so the server doesn't actually have to do the entire full join?

**Reply by jbojcic1** [reply]
Not sure how this answers the question when it doesn't work for cases when multiple rows can have the same creation time.
